sklearn/impute/_knn.py

```diff
@@ -0,0 +1,270 @@
+import numpy as np
+
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import FLOAT_DTYPES
+from ..metrics import pairwise_distances
+from ..metrics.pairwise import _NAN_METRICS
+from ..neighbors.base import _get_weights
+from ..neighbors.base import _check_weights
+from ..utils import check_array
+from ..utils import is_scalar_nan
+from ..utils.mask import _get_mask
+from ..utils.validation import check_is_fitted
+
+
+class KNNImputer(TransformerMixin, BaseEstimator):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed using values from `n_neighbors`
+    nearest neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors.
+    If a sample has more than one feature missing, then the neighbors for that
+    sample can be different depending on the particular feature being imputed.
+    When the number of available neighbors is less than `n_neighbors`, the
+    training set average for that feature is used during imputation. If a
+    feature is always missing, it is removed during `transform`.
+
+    .. versionadded:: 0.22
+
+    Parameters
+    ----------
+    missing_values : number, string, np.nan or None, default=`np.nan`
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed.
+
+    n_neighbors : int, default=5
+        Number of neighboring samples to use for imputation.
+
+    weights : str or callable, default='uniform'
+        Weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights. All points in each neighborhood are
+          weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - callable : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
```

Comment:
fwiw we could also provide median and mode weightings...

Reply:
This can be done in another PR. For now our `_get_weight` code base only support `uniform`, `distance` and callables.

Reply:
Yeah, I've realised that you can't calculate median and mode as weightings... they depend on the value, not just the weight.